Native~/main.cpp

```diff
@@ -0,0 +1,112 @@
+#include "PublicAPI.h"
+
+#include "Unity/IUnityInterface.h"
+#include "Unity/IUnityGraphics.h"
+
+#include "Logger.h"
+#include "DX12System.h"
+#include "DX12Texture.h"
+
+static IUnityInterfaces* s_UnityInterfaces = nullptr;
+static IUnityGraphics* s_Graphics = nullptr;
+
+// Unity plugin load event
+extern "C" void UNITY_INTERFACE_EXPORT UNITY_INTERFACE_API
+UnityPluginLoad(IUnityInterfaces * unityInterfaces)
+{
+    s_UnityInterfaces = unityInterfaces;
+    s_Graphics = unityInterfaces->Get<IUnityGraphics>();
+    s_Logger = std::make_unique<Logger>(unityInterfaces);
+
+    if (s_Graphics != nullptr)
+    {
+        s_Graphics->RegisterDeviceEventCallback(OnGraphicsDeviceEvent);
+    }
+
+    // Run OnGraphicsDeviceEvent(initialize) manually on plugin load
+    // to not miss the event in case the graphics device is already initialized
+    OnGraphicsDeviceEvent(kUnityGfxDeviceEventInitialize);
+}
+
+// Unity plugin unload event
+extern "C" void UNITY_INTERFACE_EXPORT UNITY_INTERFACE_API
+UnityPluginUnload()
+{
+    if (s_Graphics != nullptr)
+    {
+        s_Graphics->UnregisterDeviceEventCallback(OnGraphicsDeviceEvent);
+    }
+}
+
+static void UNITY_INTERFACE_API
+OnGraphicsDeviceEvent(UnityGfxDeviceEventType eventType)
+{
+    switch (eventType)
+    {
+        case kUnityGfxDeviceEventInitialize:
+        {
+            if (s_Graphics != nullptr && s_Graphics->GetRenderer() == kUnityGfxRendererD3D12)
+            {
+                s_DX12System = std::make_unique<DX12System>(s_UnityInterfaces);
```

Comment:
Is the variable `s_DX12System` being written to and accessed from the same thread? I think the answer might be "no". It's likely that `OnGraphicsDeviceEvent` is done on the rendering thread and `IsInitialized`, `GetD3D12Device` etc. are called from the main thread?

If not, this is technically [undefined behaviour](https://stackoverflow.com/questions/50813503/c-what-happens-when-in-one-thread-write-and-in-second-read-the-same-object-i), even though it appears to be working. In this case, you can remove the UB by putting a mutex around reads and writes to `s_DX12System`.

```cpp
static std::mutex s_DX12SystemMutex;

static void UNITY_INTERFACE_API
OnGraphicsDeviceEvent(UnityGfxDeviceEventType eventType)
{
   std::unique_lock lock(s_DX12SystemMutex);
   // initialize s_DX12System
}

extern "C" bool UNITY_INTERFACE_EXPORT
IsInitialized()
{
    std::unique_lock lock(s_DX12SystemMutex);
    return s_DX12System != nullptr && s_DX12System->IsInitialized();
}
```

Reply:
After investigation can say `OnGraphicsDeviceEvent` is always on the main thread. Luckily the plugin is pretty small and it's all single threaded. If we add `IssuePluginEvent` usage in the future we'll need to revisit this though (that one's on the render thread I'm pretty sure).